### DISCUSSION TITLE
Partial detection when using fine-tuned model for OCR.

### DISCUSSION THREAD

**Discussion by ashishupadhyaa** [mindee/doctr#1615] [Q&A]
This issue is related to #1481.

I tried the doctr model (fine-tuned) with the newest release but the issue with the fine-tuned model is that I am not able to detect the whole words present in the image but when I changed the model with pre-trained ones then the whole words present in the image are getting detected properly.

Can you guys help me out with this issue? Is it a problem with the training or code (provided in the mentioned issue at the top)? If it seems that the problem is with training then can you guys provide a proper way to train both recognition as well as detection model?

**Comment by felixdittrich92** [comment]
Hi @ashishupadhyaa :wave:,

So if it works with our pre-trained models but doesn't after fine tuning on your data it seems to be an problem with your data or training config.
To help with i need some more information :)

detection / recognition:
- would it be possible to share a code snippet how you load the fine tuned model/s and a example with the results with our pre-trained model and yours ?
- Could you share the logs from your training run ?
- Some information about your custom dataset (how many examples train / val, etc.)
- The command you used to start the run

Best regards,
Felix

**Reply by ashishupadhyaa** [reply]
- The code snippet is in issue #1481 
- The result with the provided model in the repository.
![gnome-shell-screenshot-UYCBP2](https://github.com/mindee/doctr/assets/32780193/45905998-bdc4-45e0-993b-590429e0aebb)

and in the fine-tuned model case, I won't be able to see the boxes on the right side of the image.
- Since the training was done a long time ago the logs are not there.
- I used the Pytorch framework to train the model using custom training code.

**Reply by felixdittrich92** [reply]
Hey @ashishupadhyaa :wave:,

How much data was in your train / val split ?
Unfortunately without logs it's super hard to help with.

It may also be worth considering training the model again as we have recently improved our augmentation pipeline.

And with the next release no need to remove the padding on your own anymore we have fixed this in #1627 :)

Best regards,
Felix

**Reply by ashishupadhyaa** [reply]
For detection, I had 18000 images and I split it in an 80-20 ratio for training and validation. And, for the training part I am putting more data with new symbols So once That is complete I let you know.

**Reply by ashishupadhyaa** [reply]
Hi @felixdittrich92, Is there any training script to train the detection as well as recognition model which I can use? I am trying to resolve this issue but I think I have to do the training again. Your help will be appreciated.

**Reply by felixT2K** [reply]
Hi @ashishupadhyaa 👋🏼,
Take a look here: https://mindee.github.io/doctr/using_doctr/custom_models_training.html 

:)

**Comment by Manansharma108** [comment]
i am trying doctr model = ocr_predictor(det_arch="linknet_resnet50", reco_arch="crnn_vgg16_bn", pretrained=True), and it is missing last paragraph of every page of a pdf and also missing few lines in between, it is a very simple pdf.

I am using a yolov8x custom model for bounding boxes and then the cropped boxes are being passed to doctr.

> This issue is related to #1481.
> 
> I tried the doctr model (fine-tuned) with the newest release but the issue with the fine-tuned model is that I am not able to detect the whole words present in the image but when I changed the model with pre-trained ones then the whole words present in the image are getting detected properly.
> 
> Can you guys help me out with this issue? Is it a problem with the training or code (provided in the mentioned issue at the top)? If it seems that the problem is with training then can you guys provide a proper way to train both recognition as well as detection model?

**Reply by felixdittrich92** [reply]
Hi @Manansharma108 :wave:

If you have the detection part already solved on your own and need only the recognition part you can use the standalone `recognition_predictor` by passing the already cropped words to it :)

**Reply by ashishupadhyaa** [reply]
Hi @felixdittrich92, I am getting this error while training the detection model.
![gnome-shell-screenshot-d16snn](https://github.com/user-attachments/assets/b342cb6a-2cd8-48f5-8d31-a1e9b2a17f8c)

For your information, I have checked the data and removed all the images with bboxes more than height and width of the photos, annotation pixels less than 3, and wrong annotated images. Even after that, I still get the max value of more than 1 in the `target` variable.

**Reply by felixdittrich92** [reply]
Hi @ashishupadhyaa 👋 
Yeah unfortunately there is a bug actually should be fixed next week i opened already a PR to fix this: https://github.com/mindee/doctr/pull/1715/files

CC @odulcy-mindee

**Reply by ashishupadhyaa** [reply]
Thanks for the update @felixdittrich92. I am also facing the shape issue in `parseq` model fine-tuning script:
![gnome-shell-screenshot-35moy2](https://github.com/user-attachments/assets/97662592-d5b6-401a-8c4a-8060bbbdd0a5)

**Reply by felixdittrich92** [reply]
@ashishupadhyaa Can you share the command to run ?
Do you pass the vocab correct ?

**Reply by ashishupadhyaa** [reply]
@felixdittrich92 

**command:**
`python3 references/recognition/train_pytorch.py parseq --train_path reco_data/train/ --val_path reco_data/val/ --epochs 50 --device 1 --pretrained --early-stop`

I have added one more character in the default vocab (french). I think that is giving the issue. Do you know how can I handle this? I want to add one extra character (Rupee Symbol).

**Reply by felixdittrich92** [reply]
In this case you need to create a new vocab and then pass it with `--vocab=french-extended` for example

**Reply by felixdittrich92** [reply]
https://github.com/mindee/doctr/blob/main/doctr/datasets/vocabs.py

**Reply by ashishupadhyaa** [reply]
Thanks @felixdittrich92, It is working fine.

**Reply by felixdittrich92** [reply]
> Hi @ashishupadhyaa 👋 Yeah unfortunately there is a bug actually should be fixed next week i opened already a PR to fix this: https://github.com/mindee/doctr/pull/1715/files
> 
> CC @odulcy-mindee

@ashishupadhyaa fixed on main branch now :)

**Comment by ashishupadhyaa** [comment]
Hi @felixdittrich92, I have made the changes according to your PR but when training `fast_base` detection architecture. I am facing this issue.
![gnome-shell-screenshot-1blsd7](https://github.com/user-attachments/assets/8a1c5dcb-e4f9-4b18-9495-78253ecada4a)

**Command:**

```
python3 references/detection/train_pytorch.py detect_data/train/ detect_data/val/ fast_base --epochs 200 --de
vice 1 --workers 0 --batch_size 4 --save-interval-epoch --pretrained --early-stop
```

**Reply by felixdittrich92** [reply]
Hi @ashishupadhyaa :wave:,

If i see it correct this happened after 2 epochs was already finished ?

**Reply by felixdittrich92** [reply]
I think anything strange happened while augmenting here:

Maybe cropped a part where no boxes are inside afterwards :thinking: 

```
T.OneOf([
                    T.RandomApply(T.RandomCrop(ratio=(0.6, 1.33)), 0.25),
                    T.RandomResize(scale_range=(0.4, 0.9), preserve_aspect_ratio=0.5, symmetric_pad=0.5, p=0.25),
                ]),
```

Could you try to train again and let me know if this happens a sec time ? 
If so comment the mentioned part in L280-L283 and try again.

Let me know :)

**Reply by ashishupadhyaa** [reply]
Sure @felixdittrich92, I will try and let you know.

**Reply by ashishupadhyaa** [reply]
This error occurs again after 50 epochs. I have commented out the lines and put the model on training again.
